racket: Cannot build, neiter 6.3 nor 6.2.1

## Basic info

To make sure that we are on the same page:

* Kernel: Linux yuu 4.3.6 #1-NixOS SMP Fri Feb 19 22:35:24 UTC 2016 x86_64 GNU/Linux
* System: 16.09pre78286.0bf8a1a (Emu)
* Nix version: nix-env (Nix) 1.11.2
* Nixpkgs version: 16.09pre78286.0bf8a1a

## Describe your issue here

```nix
{ stdenv, pkgs ? (import <nixpkgs> {})}:
let env = with pkgs; [ racket ];
in stdenv.mkDerivation rec {
    name = "test";
    src = ./.;
    version = "0.0.0";
    buildInputs = [ env ];
}
```

fails. Building inside `nixpkgs`: `nix-build -A racket` fails for both unstable and unstable with 00f394e2a9caab2f80b1afd6bde64d65e3ba1db7 reverted (which was the update from 6.2.1 -> 6.3)

### Expected result

`nix-shell` with `racket` build

### Actual result

Build error:

```
[...]
Copying /tmp/nix-build-racket-6.3.drv-0/racket-6.3/collects/racket/runtime-config.rkt to /tmp/nix-build-racket-6.3.drv-0/racket-6.3/src/build/racket/gc2/xform-collects/racket/runtime-config.rkt
Compiling xform support...
Done making xform-collects.
make[4]: Leaving directory '/tmp/nix-build-racket-6.3.drv-0/racket-6.3/src/build/racket/gc2'
make xobjects
make[4]: Entering directory '/tmp/nix-build-racket-6.3.drv-0/racket-6.3/src/build/racket/gc2'
env XFORM_USE_PRECOMP=xsrc/precomp.h ../racketcgc  -cqu ../../../racket/gc2/xform.rkt --setup . --depends --cpp "gcc -E -I./.. -I../../../racket/gc2/../include -pthread   -DUSE_SENORA_GC   -DMZ_USES_SHARED_LIB "  --keep-lines -o xsrc/salloc.c ../../../racket/gc2/../src/salloc.c
/nix/store/a7cagjah2n6c2d9y83lms5jj6nzavnhr-libtool-2.4.6/bin/libtool --mode=compile --tag=CC gcc -g -O2  -Wall   -I./.. -I../../../racket/gc2/../include -pthread   -DUSE_SENORA_GC   -DMZ_USES_SHARED_LIB  -c xsrc/salloc.c -o salloc.lo
libtool: compile:  gcc -g -O2 -Wall -I./.. -I../../../racket/gc2/../include -pthread -DUSE_SENORA_GC -DMZ_USES_SHARED_LIB -c xsrc/salloc.c  -fPIC -DPIC -o .libs/salloc.o
env XFORM_USE_PRECOMP=xsrc/precomp.h ../racketcgc  -cqu ../../../racket/gc2/xform.rkt --setup . --depends --cpp "gcc -E -I./.. -I../../../racket/gc2/../include -pthread   -DUSE_SENORA_GC   -DMZ_USES_SHARED_LIB "  --keep-lines -o xsrc/bignum.c ../../../racket/gc2/../src/bignum.c
/nix/store/a7cagjah2n6c2d9y83lms5jj6nzavnhr-libtool-2.4.6/bin/libtool --mode=compile --tag=CC gcc -g -O2  -Wall   -I./.. -I../../../racket/gc2/../include -pthread   -DUSE_SENORA_GC   -DMZ_USES_SHARED_LIB  -c xsrc/bignum.c -o bignum.lo
libtool: compile:  gcc -g -O2 -Wall -I./.. -I../../../racket/gc2/../include -pthread -DUSE_SENORA_GC -DMZ_USES_SHARED_LIB -c xsrc/bignum.c  -fPIC -DPIC -o .libs/bignum.o
env XFORM_USE_PRECOMP=xsrc/precomp.h ../racketcgc  -cqu ../../../racket/gc2/xform.rkt --setup . --depends --cpp "gcc -E -I./.. -I../../../racket/gc2/../include -pthread   -DUSE_SENORA_GC   -DMZ_USES_SHARED_LIB "  --keep-lines -o xsrc/bool.c ../../../racket/gc2/../src/bool.c
Error [GCING] 212 in ../../../racket/gc2/../src/bool.c: Function mz_long_double_eqv declared __xform_nongcing__, but includes a function call.
Error [GCING] 248 in ../../../racket/gc2/../src/bool.c: Function double_eqv declared __xform_nongcing__, but includes a function call.
xform: Errors converting
  context...:
   /tmp/nix-build-racket-6.3.drv-0/racket-6.3/src/build/racket/gc2/xform-collects/compiler/private/x...:8:2: xform
   /tmp/nix-build-racket-6.3.drv-0/racket-6.3/src/build/racket/gc2/xform-collects/xform/xform-mod.rkt: [running body]
Makefile:208: recipe for target 'xsrc/bool.c' failed
make[4]: *** [xsrc/bool.c] Error 1
make[4]: Leaving directory '/tmp/nix-build-racket-6.3.drv-0/racket-6.3/src/build/racket/gc2'
Makefile:172: recipe for target 'all' failed
make[3]: *** [all] Error 2
make[3]: Leaving directory '/tmp/nix-build-racket-6.3.drv-0/racket-6.3/src/build/racket/gc2'
Makefile:101: recipe for target '3m' failed
make[2]: *** [3m] Error 2
make[2]: Leaving directory '/tmp/nix-build-racket-6.3.drv-0/racket-6.3/src/build/racket'
Makefile:57: recipe for target '3m' failed
make[1]: *** [3m] Error 2
make[1]: Leaving directory '/tmp/nix-build-racket-6.3.drv-0/racket-6.3/src/build'
Makefile:51: recipe for target 'all' failed
make: *** [all] Error 2
builder for ‘/nix/store/x3ns7yr3kn88dbxa6kafqxr0v6imdsdd-racket-6.3.drv’ failed with exit code 2
error: build of ‘/nix/store/x3ns7yr3kn88dbxa6kafqxr0v6imdsdd-racket-6.3.drv’ failed
/run/current-system/sw/bin/nix-shell: failed to build all dependencies
```

### Steps to reproduce

Load the upper environment definition.

--- a/pkgs/development/interpreters/racket/default.nix
+++ b/pkgs/development/interpreters/racket/default.nix
@@ -28,6 +28,11 @@ let
     sqlite
   ];
 
+  boolPatch = fetchurl {
+    url = "http://copr-dist-git.fedorainfracloud.org/cgit/bthomas/racket/racket.git/plain/xform-errors-converting-fix.patch";
+    sha256 = "0h5g7a7w8wwj43jb8q69xldgbyxkn0y0i1na6r9fk17dd56nsm68";
+  };
+
 in
 
 stdenv.mkDerivation rec {
@@ -50,6 +55,10 @@ stdenv.mkDerivation rec {
     mkdir src/build
     cd src/build
   '';
+
+  # https://github.com/racket/racket/issues/1222
+  # Fixed upstream after the release of 6.4
+  patches = [ boolPatch ];
 
   shared = if stdenv.isDarwin then "dylib" else "shared";
   configureFlags = [ "--enable-${shared}" "--enable-lt=${libtool}/bin/libtool" ]

